File: Finder/Services/RatingService.swift
import Foundation
import SwiftUI

class RatingService: ObservableObject {
    static let shared = RatingService()

    @AppStorage("userRatingPoints") var points: Int = 0
    @AppStorage("userRatingTier") var tierRaw: Int = 1

    static let tier2Threshold = 5000

    var tier: Int {
        get { tierRaw }
        set { tierRaw = newValue }
    }

    var isTier2: Bool {
        tier >= 2
    }

    var progressToTier2: Double {
        if isTier2 { return 1.0 }
        return min(Double(points) / Double(Self.tier2Threshold), 1.0)
    }

    var pointsToNextTier: Int {
        if isTier2 { return 0 }
        return max(Self.tier2Threshold - points, 0)
    }

    private init() {
        recalculateTier()
    }

    func addPoints(_ count: Int = 1) {
        points += count
        recalculateTier()
    }

    func recalculateTier() {
        if points >= Self.tier2Threshold {
            tier = 2
        } else {
            tier = 1
        }
    }

    func resetRating() {
        points = 0
        tier = 1
    }
}

File: FinderAndroid/app/src/main/java/com/finder/app/services/RatingService.kt
package com.finder.app.services

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object RatingService {

    private const val PREFS_NAME = "finder_rating_prefs"
    private const val KEY_POINTS = "rating_points"
    private const val KEY_TIER = "rating_tier"

    const val TIER_2_THRESHOLD = 5000

    private lateinit var prefs: SharedPreferences

    private val _points = MutableStateFlow(0)
    val points: StateFlow<Int> = _points.asStateFlow()

    private val _tier = MutableStateFlow(1)
    val tier: StateFlow<Int> = _tier.asStateFlow()

    val isTier2: Boolean
        get() = _tier.value >= 2

    val progressToTier2: Float
        get() {
            if (isTier2) return 1f
            return (_points.value.toFloat() / TIER_2_THRESHOLD).coerceIn(0f, 1f)
        }

    val pointsToNextTier: Int
        get() {
            if (isTier2) return 0
            return (TIER_2_THRESHOLD - _points.value).coerceAtLeast(0)
        }

    fun init(context: Context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        _points.value = prefs.getInt(KEY_POINTS, 0)
        _tier.value = prefs.getInt(KEY_TIER, 1)
        recalculateTier()
    }

    fun addPoints(count: Int = 1) {
        _points.value += count
        recalculateTier()
        saveState()
    }

    private fun recalculateTier() {
        _tier.value = if (_points.value >= TIER_2_THRESHOLD) 2 else 1
    }

    fun setPoints(value: Int) {
        _points.value = maxOf(0, value)
        recalculateTier()
        saveState()
    }

    fun resetRating() {
        _points.value = 0
        _tier.value = 1
        saveState()
    }

    private fun saveState() {
        if (::prefs.isInitialized) {
            prefs.edit()
                .putInt(KEY_POINTS, _points.value)
                .putInt(KEY_TIER, _tier.value)
                .apply()
        }
    }
}
